In [13]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re
#import seaborn as sns

In [14]:
df = pd.read_csv('../../data/state-agency-co2e-data-2010-2014.csv')
df.head()

Organization Name  Emission Year  \
0  California Department of General Services           2013   
1  California Department of General Services           2010   
2  California Department of General Services           2010   
3  California Department of General Services           2010   
4  California Department of General Services           2010   

        Facility Name Source Name                         Activity Type  \
0  000 Fleet Vehicles         NaN      SEM: Mobile Combustion - Scope 1   
1       Lease#2243001         NaN  SEM: Purchased Electricity - Scope 2   
2       Lease#5098002         NaN  SEM: Purchased Electricity - Scope 2   
3       Lease#5107001         NaN  SEM: Purchased Electricity - Scope 2   
4       Lease#5368001         NaN  SEM: Purchased Electricity - Scope 2   

  Fuel Type Fuel End Use Sector Technology        CO2e  
0       N\A  N\A            N\A        N\A  563.025594  
1       N\A  N\A            N\A        N\A  848.793031  
2       N\A  N\A            N\A        N\A   12.768303  
3       N\A  N\A            N\A        N\A    4.683929  
4       N\A  N\A            N\A        N\A    0.265898

#### Structure

This data was stored as a csv file on the [data.ca.gov](https://data.ca.gov/dataset/calepa-state-agency-co2) website which made it very easy to read into a dataframe.

#### Granularity

In [15]:
df.shape

(8457, 10)

Each record is a different emission from different organizations, fuel type, year, activity, and end use sector. We see that data includes, for example, carbon emissions of a government building, government-owned vehicle, and even an entire fleet of government-owned vehicles. Given that there is nothing that contains a summary row, we believe that all the records capture granularity at the same level. Because the government agencies have the option to input their individual facilities, they are able to aggregate the sources at the agency level. There are 10 different features and 8,457 records. 

#### Scope

In [16]:
print('Minimum Year:', df['Emission Year'].min())
print('Maximum Year:', df['Emission Year'].max())

Minimum Year: 2010
Maximum Year: 2014


The scope of the data covers the equivalent carbon emissions from various  Government Agencies in California sourced from California EPA. 

#### Temporality

Although the scope of the data covers the 2010-2014 realm (2010-01-01 - 2014-12-31), according to the [data.ca.gov](https://data.ca.gov/dataset/calepa-state-agency-co2), this dataset was created in August 9th 2019 and was last updated in October 23, 2019.

#### Faithfulness

This data was collected by state agency reporters and entered into the Climate Registry Information System. These agencies enter in the raw data into this system which calculates the total CO2e for each source/facility. We deem this as a reliable source. However, there are a good amount of NaN values. In addition, because they are manually inputted, it would be a great idea to look at any input errors.

# Data Cleaning and Visualizations

In [24]:
df.shape

df.head()

Organization Name Emission Year  \
0  California Department of General Services          2013   
1  California Department of General Services          2010   
2  California Department of General Services          2010   
3  California Department of General Services          2010   
4  California Department of General Services          2010   

        Facility Name Source Name                         Activity Type  \
0  000 Fleet Vehicles         nan      SEM: Mobile Combustion - Scope 1   
1       Lease#2243001         nan  SEM: Purchased Electricity - Scope 2   
2       Lease#5098002         nan  SEM: Purchased Electricity - Scope 2   
3       Lease#5107001         nan  SEM: Purchased Electricity - Scope 2   
4       Lease#5368001         nan  SEM: Purchased Electricity - Scope 2   

  Fuel Type Fuel End Use Sector Technology        CO2e  
0       N\A  N\A            N\A        N\A  563.025594  
1       N\A  N\A            N\A        N\A  848.793031  
2       N\A  N\A            N\A        N\A   12.768303  
3       N\A  N\A            N\A        N\A    4.683929  
4       N\A  N\A            N\A        N\A    0.265898

In [25]:
df.columns

Index(['Organization Name', 'Emission Year', 'Facility Name', 'Source Name',
       'Activity Type', 'Fuel Type', 'Fuel', 'End Use Sector', 'Technology',
       'CO2e'],
      dtype='object')

In [36]:
df['Organization Name'].value_counts()

California Department of Fish and Wildlife (formerly California Dept of Fish and Game prior to 2013)    2272
California Department of General Services                                                                958
California Employment Development Department                                                             865
California Department of Corrections and Rehabilitation                                                  771
California Department of Forestry and Fire Protection                                                    695
California Environmental Protection Agency                                                               580
California Department of Rehabilitation                                                                  477
California Public Employees' Retirement System                                                           188
California Department of Social Services                                                                 184
California Departme

We see that some of the inputs in 'Source Name' -- particularly with the word 'Electricity' -- were inputted differently, so we look to correct this. Some reasons include typos, case-issues, etc. We also realize that some of the datatypes of the inputs are not as expected. For example, there was a random float number in the 'Fuel Type' column. Because of this, we convert all of the columns into strings and use regular expressions to in order to get the Fuel Type with a single unified catagory of "Electricity." We assume that "Purchased Electricity" is the same as "Electricity." 

In [37]:
#pd.Series.to_string(df['Fuel Type'], df['Organization Name']), df['Emission Year'], df[Source Name])

df['Fuel Type'] = df['Fuel Type'].astype(str)
df['Organization Name'] = df['Organization Name'].astype(str)
df['Emission Year'] = df['Emission Year'].astype(str)
df['Source Name'] = df['Source Name'].astype(str)


In [40]:
df['Source Name'].value_counts().head()

Electricity              1280
Purchased Electricity     892
ELECTRICITY               630
Gasoline                  568
Natural Gas               500
Name: Source Name, dtype: int64

In [39]:
df['Source Name'].value_counts()
test = pd.Series.tolist(df['Source Name'])

for element in test:
    regex = r"[Ee]+(lec)|(LEC)"
    z = re.search(regex, element)
    if z:
        print(element)

Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Purchased Electricty
Purchased Electricity
Purchased Electricty
Purchased Electricity
Pu

Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electricity
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Purchased Electricity
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Electricity
Electric
Electric
Electric
Electric
Electric
Electric
Electric
Purchased

Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Electrical Meters 1&2
Purchased Electricity
Electrical Meters 1&2
Purchased Electricity
Electrical Meters 1&2
Electricity
Purchased Electricity
Electrical Meters 1&2
Purchased Electricity
Purchased Electricity
Electrical Meters 1&2
Purchased Electricity
Purchased Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Purchased Electricity
Electricity Purchases
Electricity Purchases
Electricity Purchases
Electricity 

NameError: name 'electiity' is not defined

In [ ]:
import re

test = pd.Series.tolist(df['Source Name'])
#test = [x for x in test if x != ]
test1 = df['Source Name'].astype("string")
type(df['Source Name'])
# strinng
regex = r"([Ee]lec)"
re.findall(regex, test1)




#df['date'] = df['date'].str.replace(re_express, r'\1\2\3')

In [ ]:
wee = df[df["Organization Name"].str.contains("California Department of General Services")]
wee


In [ ]:
plt.scatter(df['Emission Year'], df['CO2e'])
plt.xlabel('Emission Years')
plt.ylabel('CO2e (Metric tons)')
plt.title("CO2e emissions between 2010-2014")
plt.show()

In [ ]:
df2010 = df[df["Emission Year"] == 2010]
df2011 = df[df["Emission Year"] == 2011]
df2012 = df[df["Emission Year"] == 2012]
df2013 = df[df["Emission Year"] == 2013]
df2014 = df[df["Emission Year"] == 2014]
df2010.columns

In [11]:
df2010['Fuel Type'].unique()

NameError: name 'df2010' is not defined

In [ ]:
plt.figure(figsize=(30,15))
plt.bar(np.array(df2010['Fuel Type'].tolist()), df2010['CO2e'])
plt.tick_params(axis="x", rotation = 45, labelsize=25)
plt.tick_params(axis = 'y', labelsize=20)
plt.ylabel('CO2e (metric tons)', fontsize = 35)
plt.xlabel('Fuel Types', fontsize = 35)
plt.title('2010 Fuel Types and CO2e Emitted', fontsize = 50)

plt.show()

In [12]:
pd.value_counts(df[''])

KeyError: ''